File: systems-for-design-and-implementation/labs/lab2/LibraryPersistance/src/library/persistance/repository/IBookRepository.java
package library.persistance.repository;


import library.model.Book;
import library.model.BookLoan;
import library.model.User;

import java.util.List;

public interface IBookRepository
{
    List<Book> getAll();

    List<Book> getAllAvailable();

    List<Book> getAllBorrowedBy(User user);

    List<Book> getAllBorrowedBy(int user_id);

    List<Book> searchByTitle(String title);

    Book getBookByID(int book_id);

    void borrowBook(Book book, User user);

    void borrowBook(int book_id, int user_id);

    void returnBook(Book book, User user);

    void returnBook(int book_id, int user_id);

    void update(Book book);

    List<BookLoan> getAllLoans();

    BookLoan getLoan(int book_id, int user_id);
}

File: systems-for-design-and-implementation/labs/lab3/LibraryPersistance/src/library/persistance/repository/IBookRepository.scala
package library.persistance.repository

import java.util

import library.model.{Book, BookLoan, User}

trait IBookRepository
{
    def getAll: util.List[Book]

    def getAllAvailable: util.List[Book]

    def getAllBorrowedBy(user: User): util.List[Book]

    def getAllBorrowedBy(user_id: Int): util.List[Book]

    def searchByTitle(title: String): util.List[Book]

    def getBookByID(book_id: Int): Book

    def borrowBook(book: Book, user: User)

    def borrowBook(book_id: Int, user_id: Int)

    def returnBook(book: Book, user: User)

    def returnBook(book_id: Int, user_id: Int)

    def update(book: Book)

    def getAllLoans: util.List[BookLoan]

    def getLoan(book_id: Int, user_id: Int): BookLoan
}

